mlir/lib/Mitigation/Transforms/MitigationMethods/Zne.cpp

```diff
@@ -0,0 +1,310 @@
+
+    Block *foldedBloc = fnFoldedOp.addEntryBlock();
+    rewriter.setInsertionPointToStart(foldedBloc);
+    TypedAttr numberQubitsAttr = rewriter.getI64IntegerAttr(numberQubits);
+    Value numberQubitsValue = rewriter.create<arith::ConstantOp>(loc, numberQubitsAttr);
+    Value allocQreg = rewriter.create<func::CallOp>(loc, fnAllocOp, numberQubitsValue).getResult(0);
+
+    Value c0 = rewriter.create<index::ConstantOp>(loc, 0);
+    Value c1 = rewriter.create<index::ConstantOp>(loc, 1);
+    int64_t sizeArgs = fnFoldedOp.getArguments().size();
+    Value size = fnFoldedOp.getArgument(sizeArgs - 1);
+    // Add scf for loop to create the folding
+    Value loopedQreg =
+        rewriter
+            .create<scf::ForOp>(
+                loc, c0, size, c1, /*iterArgsInit=*/allocQreg,
+                [&](OpBuilder &builder, Location loc, Value i, ValueRange iterArgs) {
+                    Value qreg = iterArgs.front();
+                    std::vector<Value> argsAndQreg = {fnFoldedOp.getArguments().begin(),
+                                                      fnFoldedOp.getArguments().end()};
+                    argsAndQreg.pop_back();
+                    argsAndQreg.push_back(qreg);
+
+                    // Call the function without measurements
+                    Value fnWithoutMeasurementsQreg =
+                        builder.create<func::CallOp>(loc, fnWithoutMeasurementsOp, argsAndQreg)
+                            .getResult(0);
+
+                    // Call the function without measurements in an adjoint region
+                    auto adjointOp = builder.create<quantum::AdjointOp>(loc, qregType,
+                                                                        fnWithoutMeasurementsQreg);
+                    Region *adjointRegion = &adjointOp.getRegion();
+                    Block *adjointBlock = builder.createBlock(adjointRegion, {}, qregType, loc);
+
+                    std::vector<Value> argsAndQregAdjoint = {fnFoldedOp.getArguments().begin(),
+                                                             fnFoldedOp.getArguments().end()};
+                    argsAndQregAdjoint.pop_back();
+                    argsAndQregAdjoint.push_back(adjointBlock->getArgument(0));
+                    Value fnWithoutMeasurementsAdjointQreg =
+                        builder
+                            .create<func::CallOp>(loc, fnWithoutMeasurementsOp, argsAndQregAdjoint)
+                            .getResult(0);
+                    builder.create<quantum::YieldOp>(loc, fnWithoutMeasurementsAdjointQreg);
+                    builder.setInsertionPointAfter(adjointOp);
+                    builder.create<scf::YieldOp>(loc, adjointOp.getResult());
+                })
+            .getResult(0);
+    std::vector<Value> argsAndRegMeasurement = {fnFoldedOp.getArguments().begin(),
+                                                fnFoldedOp.getArguments().end()};
+    argsAndRegMeasurement.pop_back();
```

Comment:
This can be removed. right?

Reply:
No again we want to replace the Qreg with the one from the loop